src/librustdoc/clean/cfg.rs

```diff
@@ -9,62 +9,122 @@ use std::{fmt, mem, ops};
 use rustc_session::parse::ParseSess;
-use rustc_span::Span;
 use rustc_span::symbol::{Symbol, sym};
-use {rustc_ast as ast, rustc_hir as hir};
+use rustc_span::{DUMMY_SP, Span};
 
 use crate::display::{Joined as _, MaybeDisplay, Wrapped};
 use crate::html::escape::Escape;
 
 #[cfg(test)]
 mod tests;
 
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub(crate) enum Cfg {
-    /// Accepts all configurations.
-    True,
-    /// Denies all configurations.
-    False,
-    /// A generic configuration option, e.g., `test` or `target_os = "linux"`.
-    Cfg(Symbol, Option<Symbol>),
-    /// Negates a configuration requirement, i.e., `not(x)`.
-    Not(Box<Cfg>),
-    /// Union of a list of configuration requirements, i.e., `any(...)`.
-    Any(Vec<Cfg>),
-    /// Intersection of a list of configuration requirements, i.e., `all(...)`.
-    All(Vec<Cfg>),
-}
+#[derive(Clone, Debug, Hash)]
+// Because `CfgEntry` includes `Span`, we must NEVER use `==`/`!=` operators on `Cfg` and instead
+// use `is_equivalent_to`.
+#[cfg_attr(test, derive(PartialEq))]
+pub(crate) struct Cfg(CfgEntry);
 
 #[derive(PartialEq, Debug)]
 pub(crate) struct InvalidCfgError {
     pub(crate) msg: &'static str,
     pub(crate) span: Span,
 }
 
+/// Whether the configuration consists of just `Cfg` or `Not`.
+fn is_simple_cfg(cfg: &CfgEntry) -> bool {
+    match cfg {
+        CfgEntry::Bool(..)
+        | CfgEntry::NameValue { .. }
+        | CfgEntry::Not(..)
+        | CfgEntry::Version(..) => true,
+        CfgEntry::All(..) | CfgEntry::Any(..) => false,
+    }
+}
+
+/// Whether the configuration consists of just `Cfg`, `Not` or `All`.
+fn is_all_cfg(cfg: &CfgEntry) -> bool {
+    match cfg {
+        CfgEntry::Bool(..)
+        | CfgEntry::NameValue { .. }
+        | CfgEntry::Not(..)
+        | CfgEntry::Version(..)
+        | CfgEntry::All(..) => true,
+        CfgEntry::Any(..) => false,
+    }
+}
+
+fn strip_hidden(cfg: &CfgEntry, hidden: &FxHashSet<SimpleCfg>) -> Option<CfgEntry> {
+    match cfg {
+        CfgEntry::Bool(..) => Some(cfg.clone()),
+        CfgEntry::NameValue { .. } => {
+            if !hidden.contains(&SimpleCfg::from(cfg)) {
+                Some(cfg.clone())
+            } else {
+                None
+            }
+        }
+        CfgEntry::Not(cfg, _) => {
+            if let Some(cfg) = strip_hidden(cfg, hidden) {
+                Some(CfgEntry::Not(Box::new(cfg), DUMMY_SP))
+            } else {
+                None
+            }
+        }
+        CfgEntry::Any(cfgs, _) => {
+            let cfgs =
+                cfgs.iter().filter_map(|cfg| strip_hidden(cfg, hidden)).collect::<ThinVec<_>>();
+            if cfgs.is_empty() { None } else { Some(CfgEntry::Any(cfgs, DUMMY_SP)) }
+        }
+        CfgEntry::All(cfgs, _) => {
+            let cfgs =
+                cfgs.iter().filter_map(|cfg| strip_hidden(cfg, hidden)).collect::<ThinVec<_>>();
+            if cfgs.is_empty() { None } else { Some(CfgEntry::All(cfgs, DUMMY_SP)) }
```

Comment:
Is there a reason we're converting all the spans to `DUMMY_SP` here?

Reply:
Lazyness. ^^'

We don't use the spans of the `cfg`s in rustdoc since they're only used for error reporting before rustdoc even use them.